Backend/src/main/java/com/example/acmeplex/paymentsystem/service/PaymentService.java

```diff
@@ -0,0 +1,178 @@
+        this.paymentRepository = paymentRepository;
+        this.creditRecordRepository = creditRecordRepository;
+        this.ticketRepository = ticketRepository;
+        this.registeredUserService = registeredUserService;
+    }
+
+    @Transactional
+    public double priceCalculation(List<Integer> showtimeSeats){
+        double totalPayment = 0;
+        for(Integer showtimeSeatId : showtimeSeats) {
+            double price = ticketRepository.getTicketPrice(showtimeSeatId);
+            totalPayment += price;
+        }
+        return totalPayment;
+    }
+
+    @Transactional
+    public String processMoviePayment(List<Integer> showtimeSeats, String email, String method){
+        try {
+            double totalPayment = priceCalculation(showtimeSeats);
+
+            double creditUsed=0;
+            List<CreditRecord> creditRecords = sortCreditRecords(email);
+            for(int i=0; i<creditRecords.size(); i++) {
+                CreditRecord creditRecord = creditRecords.get(i);
+                double creditPoints = creditRecord.getCreditPoints();
+                double usedPoints = creditRecord.getUsedPoints();
+                double creditAvailable = creditPoints - usedPoints;
+                if (creditAvailable > 0) {
+                    if (creditAvailable >= (totalPayment - creditUsed)) {                        
+                        creditRecord.setUsedPoints(usedPoints + totalPayment - creditUsed);
+                        creditRecordRepository.updateUsedPoints(creditRecord.getId(), creditRecord.getUsedPoints());
+                        creditUsed += creditAvailable;
+                    } else {                        
+                        creditRecord.setUsedPoints(creditPoints);
+                        creditRecordRepository.updateUsedPoints(creditRecord.getId(), creditRecord.getUsedPoints());
+                        creditUsed += creditAvailable;
+                    }
+                }
+                if (totalPayment <=creditUsed)  {
+                    break;
+                }
+            }
+
+            double remainingPayment = totalPayment - creditUsed;
+
+            int newPaymentId = paymentRepository.getLastPaymentId() + 1;
+            Payment payment = new Payment(email, method, newPaymentId, totalPayment, "ticket");
+            paymentRepository.addPayment(payment);
+
+            for(Integer showtimeSeatId : showtimeSeats) {
+                String ticketNumber = ticketRepository.getTicketNumber(showtimeSeatId);
+                paymentRepository.addPaymentTicket(payment, ticketNumber , "paid");
+            }
+
+            return "Success:"+String.valueOf(totalPayment)+" processed successfully."+String.valueOf(creditUsed)+" credit points used."+String.valueOf(totalPayment-creditUsed)+" remaining payment charged to " + method+ "card. ";        
+        } catch (RuntimeException exception) {
+            return "error: " + exception.getMessage();
```

Comment:
I just found out that returning directly will not trigger transaction rollback.
I added `TransactionAspectSupport.currentTransactionStatus().setRollbackOnly();`

Reply:
The springboot doc recommend to use Try.of from https://docs.vavr.io/#_try. 
I will use it if we are implementing failed payment part.